crates/bevy_derive/src/lib.rs

```diff
@@ -70,3 +71,8 @@ pub fn derive_app_label(input: TokenStream) -> TokenStream {
     trait_path.segments.push(format_ident!("AppLabel").into());
     derive_label(input, trait_path)
 }
+
+#[proc_macro_derive(Id)]
+pub fn derive_id(input: TokenStream) -> TokenStream {
+    id::derive_id(input)
+}
```

Comment:
Maybe make it a regular macro? Something like `define_id!(pub SystemId)` which defines the struct and add all appropriate derives.

Reply:
That's a good point! But this makes it also harder to extend those types, right? And a derive trait seems like the more idiomatic choice?

Reply:
There is not really anything that can or should be extended about those ids. Using more than one field or a non-integer field doesn't make sense for an id. Also you can still use `impl SystemId {}` of you want.

Reply:
ok, that makes sense. I will try it :) 
Thank you

Reply:
mhh I am not sure whether defining a new type inside a macro is even possible, because my generated struct is not visible across files and thus can't be imported. Any ideas on how to make the compiler know this struct before expanding the macro?

```rust

define_id!(pub MyId);


pub struct IdStruct {
    pub vis: Visibility,
    pub ident: Ident,
}

impl Parse for IdStruct {
    fn parse(input: ParseStream) -> syn::Result<Self> {
        Ok(Self {
            vis: input.parse()?,
            ident: input.parse()?,
        })
    }
}

pub fn define_id(input: TokenStream) -> TokenStream {
    let ast = parse_macro_input!(input as ItemStruct);

    let struct_name = &ast.ident;
    let visibility = &ast.vis;
    let error_message = format!("The system ran out of unique `{}`s.", struct_name);

    TokenStream::from(quote! {

        #[derive(Copy, Clone, Hash, Eq, PartialEq, Debug)]
        #visibility struct #struct_name(u64);

        impl #struct_name {
            pub fn new() -> Self {
                use std::sync::atomic::{AtomicU64, Ordering};
                static COUNTER: AtomicU64 = AtomicU64::new(0);
                COUNTER
                    .fetch_update(Ordering::Relaxed, Ordering::Relaxed, |val| {
                        val.checked_add(1)
                    })
                    .map(Self)
                    .expect(#error_message)
            }
        }
    })
}
```